File: src/main/java/io/jsql/orientstorage/adapter/OTable.java
package io.jsql.orientstorage.adapter;

import com.alibaba.druid.sql.ast.statement.SQLColumnDefinition;
import com.alibaba.druid.sql.dialect.mysql.ast.statement.MySqlCreateTableStatement;
import com.orientechnologies.orient.core.db.document.ODatabaseDocument;
import com.orientechnologies.orient.core.metadata.schema.OClass;
import com.orientechnologies.orient.core.metadata.schema.OSchema;
import com.orientechnologies.orient.core.metadata.schema.OType;
import com.orientechnologies.orient.core.record.OElement;
import io.jsql.orientstorage.sqlhander.sqlutil.MSQLutil;
import io.jsql.sql.OConnection;
import io.jsql.storage.DB;
import io.jsql.storage.StorageException;
import io.jsql.storage.Table;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import javax.annotation.PostConstruct;
import java.util.*;
import java.util.stream.Stream;

/**
 * Created by 长宏 on 2017/5/3 0003.
 */
@Service
public class OTable implements Table{
   public OTable() {
    }
    @Autowired
   private DB dbadmin;

   @PostConstruct
    void post() {
       OConnection.TABLE_ADMIN = this;
    }
    @Override
    public void droptableSyn(String dbname, String table) throws StorageException {
        ODatabaseDocument document = dbadmin.getdb(dbname);
        document.activateOnCurrentThread();
        OSchema schema = document.getMetadata().getSchema();
        if (!schema.existsClass(table)) {
            throw new StorageException("table exist");
        }
        schema.dropClass(table);
        document.close();
    }

    @Override
    public void createtableSyn(String dbname, MySqlCreateTableStatement createTableStatement) throws StorageException {

                if (!dbadmin.getallDBs().contains(dbname)) {
            throw new StorageException("db不存在");
        }
        ODatabaseDocument db = dbadmin.getdb(dbname);
        db.activateOnCurrentThread();
        OSchema oSchema = db.getMetadata().getSchema();
        String table = createTableStatement.getTableSource().toString();
        if (oSchema.existsClass(table)) {
            dbadmin.close();
            throw new StorageException("table已经存在");
        }
                OClass oClass = db
                        .createClass(table);
                oClass.setStrictMode(true);
                List<String> dstring = new ArrayList<>();
                createTableStatement.getTableElementList().forEach(a -> {
                    if (a instanceof SQLColumnDefinition) {
                        SQLColumnDefinition sqlColumnDefinition = (SQLColumnDefinition) a;
                        if (sqlColumnDefinition.getConstraints() != null) {
                            dstring.add(((SQLColumnDefinition) a).getName().toString());
                        }
                    }
                });
                Map<String, String> maps = MSQLutil.gettablenamefileds(createTableStatement);
                maps.entrySet().forEach(e -> {
                    if (e.getValue().toLowerCase().contains("int")) {
                        oClass.createProperty(e.getKey(), OType.INTEGER);
                        if (dstring.contains(e.getKey())) {
                            oClass.getProperty(e.getKey()).createIndex(OClass.INDEX_TYPE.NOTUNIQUE);
                        }
                    } else if (e.getValue().toLowerCase().contains("varchar")) {
                        oClass.createProperty(e.getKey(), OType.STRING);
                        if (dstring.contains(e.getKey())) {
                            oClass.getProperty(e.getKey()).createIndex(OClass.INDEX_TYPE.NOTUNIQUE);
                        }
                    } else if (e.getValue().toLowerCase().contains("datatime")) {
                        oClass.createProperty(e.getKey(), OType.DATETIME);
                        if (dstring.contains(e.getKey())) {
                            oClass.getProperty(e.getKey()).createIndex(OClass.INDEX_TYPE.NOTUNIQUE);
                        }
                    } else if (e.getValue().toLowerCase().contains("times")) {
                        oClass.createProperty(e.getKey(), OType.DATETIME);
                        if (dstring.contains(e.getKey())) {
                            oClass.getProperty(e.getKey()).createIndex(OClass.INDEX_TYPE.NOTUNIQUE);
                        }
                    } else {
                        oClass.createProperty(e.getKey(), OType.STRING);
                        if (dstring.contains(e.getKey())) {
                            oClass.getProperty(e.getKey()).createIndex(OClass.INDEX_TYPE.NOTUNIQUE);
                        }
                    }
                });
        dbadmin.close();
    }

    @Override
    public List<String> getalltable(String dbname) throws StorageException {
        ODatabaseDocument document = dbadmin.getdb(dbname);
        List<String> strings = new ArrayList<>();
        document.getMetadata().getSchema().getClasses().forEach(a -> strings.add(a.getName()));
        document.close();
        return strings;
    }

    @Override
    public OClass gettableclass(String tablename, String db) throws StorageException {
        ODatabaseDocument document = dbadmin.getdb(db);
        boolean b = document.getMetadata().getSchema().existsClass(tablename);
        if (!b) {
            throw new StorageException("not exits");
        }
       return document.getClass(tablename);
    }

    @Override
    public Stream<OElement> selectSyn(OClass oClass,String dbname) throws StorageException {
        return dbadmin.query("select * from " + oClass.getName(), dbname);
    }
}

File: src/main/java/io/jsql/orientstorage/adapter/OTable.kt
/*
 * Java-based distributed database like Mysql
 */

package io.jsql.orientstorage.adapter

import com.orientechnologies.orient.core.metadata.schema.OClass
import com.orientechnologies.orient.core.metadata.schema.OType
import com.orientechnologies.orient.core.record.OElement
import io.jsql.orientstorage.sqlhander.sqlutil.MSQLutil
import io.jsql.sql.OConnection
import io.jsql.storage.DB
import io.jsql.storage.StorageException
import io.jsql.storage.Table
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.*
import java.util.stream.Stream
import javax.annotation.PostConstruct

/**
 * Created by 长宏 on 2017/5/3 0003.
 */
@Service
class OTable : Table {
    @Autowired
    lateinit private var dbadmin: DB

    @PostConstruct
    internal fun post() {
        OConnection.TABLE_ADMIN = this
    }

    @Throws(StorageException::class)
    override fun droptableSyn(dbname: String, table: String) {
        val document = dbadmin.getdb(dbname)
        document.activateOnCurrentThread()
        val schema = document.metadata.schema
        if (!schema.existsClass(table)) {
            throw StorageException("table exist")
        }
        schema.dropClass(table)
        dbadmin.close(document)
    }

    @Throws(StorageException::class)
    override fun createtableSyn(dbname: String, createTableStatement: MySqlCreateTableStatement) {

        if (!dbadmin.getallDBs().contains(dbname)) {
            throw StorageException("db不存在")
        }
        val db = dbadmin.getdb(dbname)
        db.activateOnCurrentThread()
        val oSchema = db.metadata.schema
        val table = createTableStatement.tableSource.toString()
        if (oSchema.existsClass(table)) {
            dbadmin.close(db)
            throw StorageException("table已经存在")
        }
        val oClass = db
                .createClass(table)
        oClass.isStrictMode = true
        val dstring = ArrayList<String>()
        createTableStatement.tableElementList.forEach { a ->
            if (a is SQLColumnDefinition) {
                if (a.constraints != null) {
                    dstring.add(a.name.toString())
                }
            }
        }
        val maps = MSQLutil.gettablenamefileds(createTableStatement)
        maps.entries.forEach { e ->
            if (e.value.toLowerCase().contains("int")) {
                oClass.createProperty(e.key, OType.INTEGER)
                if (dstring.contains(e.key)) {
                    oClass.getProperty(e.key).createIndex(OClass.INDEX_TYPE.NOTUNIQUE)
                }
            } else if (e.value.toLowerCase().contains("varchar")) {
                oClass.createProperty(e.key, OType.STRING)
                if (dstring.contains(e.key)) {
                    oClass.getProperty(e.key).createIndex(OClass.INDEX_TYPE.NOTUNIQUE)
                }
            } else if (e.value.toLowerCase().contains("datetime")) {
                oClass.createProperty(e.key, OType.DATETIME)
                if (dstring.contains(e.key)) {
                    oClass.getProperty(e.key).createIndex(OClass.INDEX_TYPE.NOTUNIQUE)
                }
            } else if (e.value.toLowerCase().contains("times")) {
                oClass.createProperty(e.key, OType.DATETIME)
                if (dstring.contains(e.key)) {
                    oClass.getProperty(e.key).createIndex(OClass.INDEX_TYPE.NOTUNIQUE)
                }
            } else {
                oClass.createProperty(e.key, OType.STRING)
                if (dstring.contains(e.key)) {
                    oClass.getProperty(e.key).createIndex(OClass.INDEX_TYPE.NOTUNIQUE)
                }
            }
        }
        dbadmin.close(db)
    }

    @Throws(StorageException::class)
    override fun getalltable(dbname: String): List<String> {
        val document = dbadmin.getdb(dbname)
        document.activateOnCurrentThread()
        val strings = ArrayList<String>()
        document.metadata.schema.classes.forEach { a -> strings.add(a.name) }
        dbadmin.close(document)
        return strings
    }

    @Throws(StorageException::class)
    override fun gettableclass(tablename: String, db: String): OClass {
        val document = dbadmin.getdb(db)
        val b = document.metadata.schema.existsClass(tablename)
        if (!b) {
            dbadmin.close(document)
            throw StorageException("not exits")
        }
        val aClass = document.getClass(tablename)
        dbadmin.close(document)
        return aClass
    }

    @Throws(StorageException::class)
    override fun selectSyn(oClass: OClass, dbname: String): Stream<OElement> {
        return dbadmin.query("select * from " + oClass.name, dbname)
    }
}
